app/controllers/fhir/r4/api_controller.rb

```diff
@@ -424,6 +424,120 @@ def all
     status_ok(bundle) && return
   end
 
+  # Kick off async generation of bulk FHIR data for all accessible patients
+  #
+  # GET /fhir/r4/Patient/$export
+  def bulk_data_export
+    # Require all scopes for all possible resources being returned
+    return if doorkeeper_authorize!(*PATIENT_READ_SCOPES)
+    return if doorkeeper_authorize!(*OBSERVATION_READ_SCOPES)
+    return if doorkeeper_authorize!(*QUESTIONNAIRE_RESPONSE_READ_SCOPES)
+    return if doorkeeper_authorize!(*RELATED_PERSON_READ_SCOPES)
+    return if doorkeeper_authorize!(*IMMUNIZATION_READ_SCOPES)
+    return if doorkeeper_authorize!(*PROVENANCE_READ_SCOPES)
+
+    status_not_acceptable_with_custom_errors(["'Accept' header must have a value of 'application/fhir+json'"]) && return unless accept_header?
+
+    unless request.headers['Prefer']&.include?('respond-async')
+      status_unprocessable_entity_with_custom_errors(["'Prefer' header must have value of 'respond-async'"],
+                                                     '') && return
+    end
+
+    unless @m2m_workflow
+      status_unauthorized_with_custom_errors(['Bulk export requires a client application registered for the Backend Services Workflow']) && return
+    end
+
+    client_app = current_client_application
+    if client_app.exported_recently?
+      status_too_many_requests_with_custom_errors(['Client already initiated an export of this type in the last 15 minutes. Please try again later']) && return
+    end
+
+    begin
+      since = params.permit(:_since)[:_since]
+      since = DateTime.parse(since) unless since.nil?
+    rescue Date::Error
+      status_unprocessable_entity_with_custom_errors(['Invalid Date in _since parameter. Please use the FHIR instant datatype'], '') && return
+    end
+
+    # Create the download to uniquely id this request, and queue the ExportFhirJob
+    download = ApiDownload.create(application_id: client_app.id, url: request.url)
+    export_job = ExportFhirJob.perform_later(client_app, download, { since: since })
+    # Add the job_id of the job to the download for reference in status requests
+    download.update(job_id: export_job.provider_job_id)
+
+    respond_to do |format|
+      format.any { head(:accepted, content_location: "#{root_url}fhir/r4/ExportStatus/#{download.id}") }
+    end
+  end
+
+  # Give an update on the status of a bulk FHIR data request
+  #
+  # GET /fhir/r4/ExportStatus/[:id]
+  def export_status
+    id = params.permit(:id)[:id]
+    download = ApiDownload.find_by_id(id)
+
+    case Sidekiq::Status.status(download&.job_id)
+    when :complete
+      response_json = {
+        transactionTime: Sidekiq::Status.get(download.job_id, :transaction_time),
```

Comment:
From what it looks like Status.get may return a Redis response code string, and not the value and that case should be handled. [code](https://github.com/kenaniah/sidekiq-status/blob/1366fb6647431606744702f5419407a5b483c141/lib/sidekiq-status/storage.rb#L68)

Reply:
In the `ExportFhirJob`, it's already transformed into a string (following the instance regex) [here](https://github.com/SaraAlert/SaraAlert/blob/19d5016a3b6b4e37c1db94ce07a9891bd9231646/app/jobs/export_fhir_job.rb#L13), so I think that should be good.

Ah good catch. I'm thinking if we can get the `transaction_time` correctly, it is just an error and we say we can't return the status. Since `transactionTime` is a required field (and also an important field), I don't think we should give status updates if we can't get it. What do you think?

Reply:
Yes if transactionTime isn't a DateTime then return 500.

Reply:
Added [here](https://github.com/SaraAlert/SaraAlert/pull/987/commits/6977cd314b84a7c5ecfd8d2d2ac200bf5d218e7b#diff-e79e38176ee697a8541cffba23c60d9decaa69382ca5c6e875da033508e5b428R489-R494)